File: creational/singleton/singleton.js
// Singleton Pattern in JavaScript
// Mục tiêu: Chỉ có duy nhất 1 instance của class được tạo trong toàn bộ ứng dụng

// Cách 1: Sử dụng Class với static
class Singleton {
    constructor() {
        if (Singleton.instance) {
            return Singleton.instance;
        }
        Singleton.instance = this;
        this.value = null;
    }
}

// Cách 2: Thực tế - Database connection singleton
class Database {
    constructor() {
        if (Database.instance) {
            return Database.instance;
        }
        
        this.connection = `Connected to DB at ${Date.now()}`;
        Database.instance = this;
    }
    
    query(sql) {
        return `Executing: ${sql} on ${this.connection}`;
    }
}

// Test
const db1 = new Database();
const db2 = new Database();

console.log(`db1 ID: ${db1.constructor.name}`);
console.log(`db2 ID: ${db2.constructor.name}`);
console.log(`Cùng object? ${db1 === db2}`); // true
console.log(db1.query("SELECT * FROM users"));
console.log(db2.query("SELECT * FROM products"));

// Module export
module.exports = Database;

File: creational/singleton/singleton.py
# Singleton Pattern in Python
# Mục tiêu: Chỉ có duy nhất 1 instance của class được tạo trong toàn bộ ứng dụng

class Singleton:
    """
    Cách 1: Sử dụng Metaclass (Pythonic)
    """
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance
    
    def __init__(self):
        self.value = None


class Database:
    """
    Ví dụ thực tế: Database connection singleton
    """
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if self._initialized:
            return
        self.connection = f"Connected to DB at {id(self)}"
        self._initialized = True
    
    def query(self, sql):
        return f"Executing: {sql} on {self.connection}"


# Test
if __name__ == "__main__":
    db1 = Database()
    db2 = Database()
    
    print(f"db1 ID: {id(db1)}")
    print(f"db2 ID: {id(db2)}")
    print(f"Cùng object? {db1 is db2}")  # True
    print(db1.query("SELECT * FROM users"))
    print(db2.query("SELECT * FROM products"))
